### DISCUSSION TITLE
[Bug]: Storybook Actions not working in Angular projects without Compodoc

### DISCUSSION THREAD

**Discussion by jan-teschner-webcom** [storybookjs/storybook#29138] [Help] [answered]
### Describe the bug

When installing Storybook into an Angular project, Storybook Actions don't work unless Compodoc is installed alongside Storybook.

### Reproduction link

https://github.com/jan-teschner-webcom/angular-storybook-error

### Reproduction steps

1. Create a new angular project using `ng new project-name`. Choose default values stylesheet (CSS) and SSR (No)
2. Install storybook in that project using `npx storybook@latest init`. Choose No for Compodoc. (The reproduction link above contains a repository with these first two steps already completed)
3. With Storybook open, click on the "Primary" story of the "Button" component.
4. Clicking on the button does not create click events in the actions tab

5. (optionally) Create another angular project, add storybook and choose Yes for Compodoc. You will find that the click events are logged as expected.

### System

```bash
System:
    OS: macOS 14.6.1
    CPU: (10) arm64 Apple M1 Max
    Shell: 5.9 - /bin/zsh
  Binaries:
    Node: 20.16.0 - ~/.nvm/versions/node/v20.16.0/bin/node
    npm: 10.8.1 - ~/.nvm/versions/node/v20.16.0/bin/npm <----- active
  Browsers:
    Chrome: 128.0.6613.137
    Safari: 17.6
  npmPackages:
    @storybook/addon-essentials: ^8.3.0 => 8.3.0 
    @storybook/addon-interactions: ^8.3.0 => 8.3.0 
    @storybook/addon-links: ^8.3.0 => 8.3.0 
    @storybook/addon-onboarding: ^8.3.0 => 8.3.0 
    @storybook/angular: ^8.3.0 => 8.3.0 
    @storybook/blocks: ^8.3.0 => 8.3.0 
    @storybook/test: ^8.3.0 => 8.3.0 
    storybook: ^8.3.0 => 8.3.0
```

### Additional context

Another user had a similar problem and created a discussion here: https://github.com/storybookjs/storybook/discussions/26867
They were able to install Storybook 7.6.17 and then upgrade to 8 via the command line to circumvent the problem. For me this is not possible, as Storybook 7.6.17 is not compatible with my Angular version (18.1.0). Going from Storybook 7.6.20 to 8 also doesn't work, since the issue also exists in Storybook 7.6.20.
Also, I don't want to install Compodoc, so that's not a viable workaround either.

**Comment by valentinpalkovic** [comment]
If the arguments/properties of a component can't be inferred or aren't set manually, the `actions: { argTypesRegex: '^on.*' },` pattern in your preview.js can't be applied.

Please follow the following documentation for more context:
https://storybook.js.org/docs/essentials/actions#via-storybooktest-fn-spy-function

(make sure to select "Angular" as framework at the top of the page)

**Reply by jan-teschner-webcom** [reply]
As I said in the original question, this is a completely fresh default installation of Angular and Storybook. There is no preview.js and in preview.ts, there is no argTypesRegex. The default installation uses the same args: { onClick: fn() } as specified in the link you sent, but neither that, nor an argTypesRegex work.
As far as I can tell, the only way to fix it is to install Compodoc during setup, which seems like a bug.

**Comment by Marklb** [comment, marked as answer]
This is something that probably needs to be fixed, because the docs are a bit misleading and I think it should work as described in the docs. This looks related to what I found for someone with a different problem, but caused by that same code. https://github.com/storybookjs/storybook/discussions/29050#discussioncomment-10560749

At the moment, Storybook only adds an action for a story that manually specifies `action: 'something'` in the properties `argTypes` or if the property is recognized as an output in the Compodoc json.

So, if `onClick` is not an output in the Compodoc json, which is obviously the case if you don't use Compodoc, then the first example from the docs would have to be like the following:

```ts
import type { Meta, StoryObj } from '@storybook/angular';
import { fn } from '@storybook/test';

import { Button } from './button.component';

const meta: Meta<Button> = {
  component: Button,
  // The `fn` spy is not detected as an indication that this should be an action.
  // args: { onClick: fn() },
  argTypes: {
    // This should be added for you, but currently the outputs in the Compodoc json
    // are all that it does this for.
    onClick: { action: 'onClick' },
  },
};

export default meta;
```

**Reply by jan-teschner-webcom** [reply]
Thank you Mark, this does indeed work! The actions are now properly logged in the Actions tab. I agree that this definitely has to at least be included in the docs, because so far nothing suggests that this functionality requires Storybook to be installed.

However, it still doesn't seem to be configured correctly, because sometimes there are two events added to the Actions tab! The second one is the correct event, but the first one is the same event wrapped in an array. I don't know if it's expected functionality or another bug, but it seems somewhat inconsistent. One way to reproduce it is to switch to the Primary Button Story, then Secondary, then Primary again, then Secondary and then click the button:
<img width="494" alt="image" src="https://github.com/user-attachments/assets/6e325e21-2c17-47e4-8f57-274410297d0c">
